File: apps/api/src/domain/identity/value-objects/Role.ts
export enum RoleScope {
  ADMIN = 'admin',
  PRODUCT = 'product',
}

export class Role {
  private constructor(
    public readonly code: string,
    public readonly scope: RoleScope,
  ) {}

  static readonly OWNER = new Role('owner', RoleScope.ADMIN);
  static readonly ASSISTANT = new Role('assistant', RoleScope.ADMIN);
  static readonly EDITOR = new Role('editor', RoleScope.ADMIN);
  static readonly CLIENT = new Role('client', RoleScope.PRODUCT);

  static fromCode(code: string): Role {
    switch (code) {
      case 'owner':
        return Role.OWNER;
      case 'assistant':
        return Role.ASSISTANT;
      case 'editor':
        return Role.EDITOR;
      case 'client':
        return Role.CLIENT;
      default:
        throw new Error(`Unknown role code: ${code}`);
    }
  }

  equals(other: Role): boolean {
    return this.code === other.code && this.scope === other.scope;
  }

  isAdmin(): boolean {
    return this.scope === RoleScope.ADMIN;
  }
}

File: backend/domain/identity/value_objects/role.py
"""
Role Value Object.
"""
from domain.shared.value_object import ValueObject


class Role(ValueObject):
    """Value Object для роли пользователя.
    
    Роли:
    - OWNER: Владелец (психолог) - полный доступ
    - ASSISTANT: Ассистент - расписание/лиды/модерация
    - EDITOR: Редактор - контент/ресурсы
    - CLIENT: Клиент - базовый доступ
    """
    
    def __init__(self, code: str, scope: str):
        if code not in ['owner', 'assistant', 'editor', 'client']:
            raise ValueError(f"Invalid role code: {code}")
        if scope not in ['admin', 'product']:
            raise ValueError(f"Invalid role scope: {scope}")
        
        self._code = code
        self._scope = scope
    
    @property
    def code(self) -> str:
        return self._code
    
    @property
    def scope(self) -> str:
        return self._scope
    
    def is_admin(self) -> bool:
        """Проверяет, является ли роль административной."""
        return self._scope == 'admin'
    
    # Предопределенные роли
    OWNER = None
    ASSISTANT = None
    EDITOR = None
    CLIENT = None


# Инициализация предопределенных ролей
Role.OWNER = Role('owner', 'admin')
Role.ASSISTANT = Role('assistant', 'admin')
Role.EDITOR = Role('editor', 'admin')
Role.CLIENT = Role('client', 'product')
